docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+ * Intersections: a type that must implement all types, ex. `(A & imm)`.
+
+Capabilities offer language and runtime guarantees:
+ * `iso`, or `isolated`, means this is a pointer to an exclusive area of memory that only this pointer has access to in the entire program.
+ * `mut`, or `mutable`, means this object can be changed at will, which means it must be inside an area of memory you have exclusive access to, ie. traced back to an `iso` you hold the rights.
+ * `imm`, or `immutable`, means no one can write to, so anyone can read from at any time. You can make `iso` regions immutable but not the other way round, as there's no guarantees who's still reading from it.
+ * `readonly` is a local capability that is used when you want to make sure an external function can't modify, but you still hold a mutable reference.
+
+Note that type capabilities, unions and intersections are types themselves.
+So `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+You can freeze an `iso` region into `imm`, but unfreeze an `imm` region.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific non-trivial runtime semantics.
+
+The type system guarantees that:
+ * Structural soundness: Interfaces and classes form valid concrete types, operations and functions take in the correct (sometimes inferred) types.
+ * Capability guarantees are followed: The compiler checks operations on types, the runtime guarantees the semantics is followed.
+ * Runtime safety: A program that compiles is a program that runs safely, without crashes, traps or unhandled exceptions, including asynchronous behaviour, but not necessarily producing the right result or doing what you expect.
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can convert the object to a string representation
+interface Serialisable
+{
+  // All derived classes will have the common field
+  var common : F64;
+
+  // Derived classes must implement this method
+  toString() : String;
+};
+
+class MyType : Serialisable
+{
+  // Inherits `common`
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `toString`
+  toString() : String
+  {
+    ...
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
+
+They can also be a private member, if its name starts with `_`.
```

Comment:
I would change this to

"Fields whose name starts with `_` are private"

(but then what private means for verona needs to be defined somewhere, perhaps here.)

Reply:
I believe the closest we came to consensus was that we have two kinds of visibility:

 - Public (anything can read / write the field or call the method)
 - Package (anything within the same package can read / write the field or call the method).

Protected makes little sense when subclassing is distinct from subtyping and private is not very important when you have a sensible module system, because it can be enforced by convention (and Package avoids the need for `friend` and similar).

Reply:
@SeanTAllen feel free to suggest better wording as a patch.

@davidchisnall I thought that, too, but then when talking about names this week, the `_` naming was mentioned to allow "private variables". This was a paragraph that was meant to raise discussion.

I'd be happy with no `_` rules and everything is package-private. But I'd also be happy with class-private via `_` in addition to package-private.

Reply:
I was being lazy in the discussion and saying private when I meant package, because I fundamentally don't believe sub-package private actually exists.